### ISSUE TITLE
Piping samtools view and sort causes error [main_samview] random alignment retrieval only works for indexed BAM or CRAM files.

### ISSUE THREAD

**Issue by rbenel** [samtools/samtools#1072]
I am using samtools1.9, on my local computer as well as on our Cluster. 

Prior to running on the Cluster, I ran a single file locally using this command
`samtools view -b ./path/to/.sam | samtools sort -o ./path/to/output.bam` with no problems. 

However, on the cluster I have tried to same command as well as added multiple flags, but I still receive this error/warning message every time, "[main_samview] random alignment retrieval only works for indexed BAM or CRAM files."

I tried adding `-u`, `-S` `-h` to samtools view and a `-` after the .bam file and I still have the same issue.. 

Thanks!

**Comment by daviesrob** [comment]
`samtools sort` has been able to read SAM (and BAM and CRAM) files for quite a long time now, so you can simplify your command to:

```
samtools sort -o ./path/to/output.bam ./path/to/input.sam
```

Also, if running on a machine with plenty of memory, you should use the `-m` option to increase the amount that sort can use.

You haven't said exactly what command you're running on your cluster, but `samtools view` thinks that there is a region specifier after the name of your input file.  I'm guessing that your pipeline isn't being run under a normal shell so the `|` symbol is somehow ending up on samtools' command line but without more information I can only speculate on how that might happen.  It would be best to consult a local systems administrator who knows how your cluster is set up and works.

**Comment by rbenel** [comment]
Hi @daviesrob, 

I am happy to hear that `samtools sort` can read SAM files as an input. I have just adjusted the script to include only `sort`. 

I apologize for not specifying, I am using a cluster which uses the `qsub` system, so there is a wrapper file that feeds single files to the `samtools` script, so that that process can run in parallel. It is definitely possible that this is the issue. 

Now that I have removed the `|` I will update if it is solved! 

Thanks again!
